controlplane/pkg/nsm/endpoint_client.go

```diff
@@ -19,7 +20,8 @@ func (c *endpointClient) Request(ctx context.Context, request nsm.NSMRequest) (n
 	if c == nil || c.client == nil {
 		return nil, fmt.Errorf("NSE Connection is not initialized...")
 	}
-	return c.client.Request(ctx, request.(*networkservice.NetworkServiceRequest))
+	response, err := c.client.Request(ctx, request.(*networkservice.NetworkServiceRequest))
+	return proto.Clone(response).(*connection.Connection), err
```

Comment:
I don't quite follow why we need to clone here.  Could you explain a bit more?

Reply:
```go
func (srv *networkServiceManager) performNSERequest(requestId string, ctx context.Context, endpoint *registry.NSERegistration, requestConnection nsm.NSMConnection, request nsm.NSMRequest, dp *model.Dataplane, existingConnection *model.ClientConnection) (*model.ClientConnection, error) {

    ...

    logrus.Infof("NSM:(7.2.6.2-%v) Requesting NSE with request %v", requestId, message) 
    nseConnection, e := client.Request(ctx, message)

    ...

    // 7.2.6.2.4 update connection parameters, add workspace if local nse
    srv.updateConnectionParameters(requestId, nseConnection, endpoint)

    ...
}
```

Method `client.Request(...)` returns struct `nseConnection` and we start modify that struct in `updateConnectionParameters`:

```go
func (srv *networkServiceManager) updateConnectionParameters(requestId string, nseConnection nsm.NSMConnection, endpoint *registry.NSERegistration) {
	if srv.isLocalEndpoint(endpoint) {
		modelEp := srv.model.GetEndpoint(endpoint.GetNetworkserviceEndpoint().GetEndpointName())
		if modelEp != nil { // In case of tests this could be empty
			nseConnection.(*connection.Connection).GetMechanism().GetParameters()[connection.Workspace] = modelEp.Workspace
			nseConnection.(*connection.Connection).GetMechanism().GetParameters()[connection.WorkspaceNSEName] = modelEp.Endpoint.NetworkserviceEndpoint.EndpointName
		}
		logrus.Infof("NSM:(7.2.6.2.4-%v) Update Local NSE connection parameters: %v", requestId, nseConnection.(*connection.Connection).GetMechanism())
	}
}
```

But according to logs, when `client.Request` returns `nseConnection` to us, it also pass it to `jaeger-client-go` which is working in separate goroutine. In logs we can see that `jaeger-client-go` calls `nseConnection.String()` which starts iterating over `Mechanism.Parameters` map. That's why sometimes we can catch `fatal error: concurrent map iteration and map write`. 

You can check out logs in Issue https://github.com/networkservicemesh/networkservicemesh/issues/901

Reply:
Nice catch! :)